core/embed/rust/src/ui/model_tr/layout.rs

```diff
@@ -108,6 +159,146 @@ extern "C" fn new_confirm_text(n_args: usize, args: *const Obj, kwargs: *mut Map
     unsafe { util::try_with_args_and_kwargs(n_args, args, kwargs, block) }
 }
 
+extern "C" fn request_pin(n_args: usize, args: *const Obj, kwargs: *mut Map) -> Obj {
+    let block = |_args: &[Obj], kwargs: &Map| {
+        let prompt: StrBuffer = kwargs.get(Qstr::MP_QSTR_prompt)?.try_into()?;
+        let subprompt: StrBuffer = kwargs.get(Qstr::MP_QSTR_subprompt)?.try_into()?;
+        let allow_cancel: Option<bool> =
+            kwargs.get(Qstr::MP_QSTR_allow_cancel)?.try_into_option()?;
+        let shuffle: Option<bool> = kwargs.get(Qstr::MP_QSTR_shuffle)?.try_into_option()?;
+
+        let obj = LayoutObj::new(
+            PinPage::new(
+                prompt,
+                subprompt,
+                allow_cancel.unwrap_or(true),
+                shuffle.unwrap_or(false),
+            )
+            .into_child(),
+        )?;
+        Ok(obj.into())
+    };
+    unsafe { util::try_with_args_and_kwargs(n_args, args, kwargs, block) }
+}
+
+extern "C" fn show_share_words(n_args: usize, args: *const Obj, kwargs: *mut Map) -> Obj {
+    let block = |_args: &[Obj], kwargs: &Map| {
+        let share_words: StrBuffer = kwargs.get(Qstr::MP_QSTR_share_words)?.try_into()?;
+        let title = "Recovery seed";
+
+        let share_words_len = share_words.split(',').count();
+        let share_words_len_str: String<10> = String::from(share_words_len as i16);
+
+        let beginning_text: String<50> = util::format_string([
```

Comment:
fwiw this might be a usecase for `FormattedText`

Reply:
`FormattedText` are then not compatible with `Paragraphs` - but maybe there is a better how to display them